crates/nri-resctrl-plugin/src/lib.rs

```diff
@@ -155,6 +232,111 @@ impl<P: FsProvider> ResctrlPlugin<P> {
             warn!("resctrl-plugin: failed to send event: {}", e);
         }
     }
+
+    // Create or fetch pod state and ensure group exists
+    fn handle_new_pod(&self, pod: &nri::api::PodSandbox) -> Option<String> {
+        let pod_uid = &pod.uid;
+        let mut st = self.state.lock().unwrap();
+        let ps = st.pods.entry(pod_uid.clone()).or_insert(PodState {
+            group_path: None,
+            total_containers: 0,
+            reconciled_containers: 0,
+        });
+        if ps.group_path.is_none() {
+            match self.resctrl.create_group(pod_uid) {
+                Ok(p) => {
+                    ps.group_path = Some(p.clone());
+                    Some(p)
+                }
+                Err(e) => {
+                    warn!("resctrl-plugin: failed to create group for pod {}: {}", pod_uid, e);
+                    None
+                }
+            }
+        } else {
+            ps.group_path.clone()
+        }
+    }
+
+    fn handle_new_container(&self, pod: &nri::api::PodSandbox, container: &nri::api::Container) {
+        let pod_uid = pod.uid.clone();
+        let group_path = match self.handle_new_pod(pod) {
+            Some(g) => g,
+            None => {
+                // Emit failure for pod without group
+                let ev = PodResctrlEvent::Added(PodResctrlAdded {
+                    pod_uid: pod_uid.clone(),
+                    group_path: None,
+                    state: AssignmentState::Failure,
+                    total_containers: 0,
+                    reconciled_containers: 0,
+                });
+                self.emit_event(ev);
+                return;
+            }
+        };
+
+        // Enumerate container PIDs and reconcile just this container into the pod group
+        let passes = self.cfg.max_reconcile_passes;
+        let pids = self.pid_source.pids_for_container(container);
+        let res = self
+            .resctrl
+            .reconcile_group(&group_path, || Ok(pids.clone()), passes);
```

Comment:
we should pass a function that resolves the pid on every call, not use a static set like here